File: backend/simbrief_client.py
"""Fetches a pilot's last-generated SimBrief flight plan (OFP) - the public,
key-less read endpoint (not the dispatch/generation API, which needs a
registered key). No auth beyond the user's own SimBrief/Navigraph username.
"""
import requests

SIMBRIEF_URL = "https://www.simbrief.com/api/xml.fetcher.php"
_TIMEOUT_S = 8


def _parse_ofp(data):
    """Pure parsing of an already-decoded SimBrief OFP JSON payload - kept
    separate from fetch_latest_ofp below so it's testable without a network
    call (same split as update_check.py's _parse_version/check_for_update).

    origin/destination/alternate are load-bearing for the scenery-matching
    feature, so a missing one raises (caller turns that into an error).
    duration_minutes/aircraft_name/planned_at are a best-effort summary so
    the UI can show the user at a glance what plan they're looking at and
    when it was generated (SimBrief has no concept of "current" flight -
    it's just whatever was last generated, could be stale) - any of those
    being absent/malformed just comes back None, never raises."""
    origin = data["origin"]["icao_code"]
    destination = data["destination"]["icao_code"]
    alternate = (data.get("alternate") or {}).get("icao_code") or None

    times = data.get("times") or {}
    aircraft = data.get("aircraft") or {}
    params = data.get("params") or {}

    duration_minutes = None
    try:
        duration_minutes = round(int(times.get("est_time_enroute")) / 60)
    except (TypeError, ValueError):
        pass

    planned_at = None
    try:
        planned_at = int(params.get("time_generated"))
    except (TypeError, ValueError):
        pass

    return {
        "origin": origin,
        "destination": destination,
        "alternate": alternate,
        "duration_minutes": duration_minutes,
        "aircraft_name": aircraft.get("name") or None,
        "planned_at": planned_at,
        "route_points": _parse_route_points(data),
    }


def _parse_route_points(data):
    """The actual planned route (origin -> destination only - SimBrief does
    not compute a navlog for the alternate diversion, so that leg stays a
    straight line on the map same as before) as an ordered list of
    {"lat": float, "lon": float} - lets the map draw the real route instead
    of a straight line between airports. Best-effort: a missing/malformed
    navlog just means an empty list, the map falls back to a straight line
    - never treated as fatal, unlike origin/destination above."""
    points = []
    for fix in (data.get("navlog") or {}).get("fix") or []:
        try:
            points.append({"lat": float(fix["pos_lat"]), "lon": float(fix["pos_long"])})
        except (TypeError, ValueError, KeyError):
            continue
    return points


def fetch_latest_ofp(username):
    """Returns {"ok": True, "origin": "KJFK", "destination": "KLAX",
    "alternate": "KONT"|None, "duration_minutes": int|None,
    "aircraft_name": str|None, "planned_at": int|None (unix seconds),
    "route_points": [{"lat": float, "lon": float}, ...]} or
    {"ok": False, "error": <str>}. The error string is a translation key
    when recognized, else a raw message."""
    if not username or not username.strip():
        return {"ok": False, "error": "simbrief_no_username"}

    try:
        response = requests.get(
            SIMBRIEF_URL, params={"username": username.strip(), "json": 1}, timeout=_TIMEOUT_S
        )
    except requests.RequestException as e:
        return {"ok": False, "error": str(e)}

    if response.status_code != 200:
        return {"ok": False, "error": f"HTTP {response.status_code}"}

    try:
        parsed = _parse_ofp(response.json())
    except Exception as e:
        return {"ok": False, "error": f"unexpected response ({e})"}

    return {"ok": True, **parsed}

File: dotnet/src/FlightOpsHub.Core/SimbriefClient.cs
using System.Net.Http;
using System.Text.Json.Nodes;

namespace FlightOpsHub.Core;

/// <summary>
/// Port of backend/simbrief_client.py - fetches a pilot's last-generated
/// SimBrief flight plan (OFP) via the public, key-less read endpoint.
/// </summary>
public static class SimbriefClient
{
    private const string SimbriefUrl = "https://www.simbrief.com/api/xml.fetcher.php";
    private static readonly TimeSpan Timeout = TimeSpan.FromSeconds(8);

    private static readonly HttpClient Http = new() { Timeout = Timeout };

    /// <summary>
    /// Pure parsing of an already-decoded SimBrief OFP JSON payload - kept
    /// separate from FetchLatestOfp so it's testable without a network
    /// call. origin/destination are load-bearing for the scenery-matching
    /// feature, so a missing one throws (caller turns that into an error).
    /// duration_minutes/aircraft_name/planned_at are a best-effort summary;
    /// any of those being absent/malformed just comes back null, never throws.
    /// </summary>
    public static JsonObject ParseOfp(JsonObject data)
    {
        var origin = GetRequiredIcao(data, "origin");
        var destination = GetRequiredIcao(data, "destination");
        var alternate = data["alternate"] is JsonObject altObj ? GetString(altObj, "icao_code") : null;

        var times = data["times"] as JsonObject ?? new JsonObject();
        var aircraft = data["aircraft"] as JsonObject ?? new JsonObject();
        var parameters = data["params"] as JsonObject ?? new JsonObject();

        long? durationMinutes = null;
        if (TryGetLong(times["est_time_enroute"], out var enrouteSeconds))
        {
            durationMinutes = (long)Math.Round(enrouteSeconds / 60.0);
        }

        long? plannedAt = TryGetLong(parameters["time_generated"], out var generated) ? generated : null;

        return new JsonObject
        {
            ["origin"] = origin,
            ["destination"] = destination,
            ["alternate"] = string.IsNullOrEmpty(alternate) ? null : alternate,
            ["duration_minutes"] = durationMinutes,
            ["aircraft_name"] = string.IsNullOrEmpty(GetString(aircraft, "name")) ? null : GetString(aircraft, "name"),
            ["planned_at"] = plannedAt,
            ["route_points"] = ParseRoutePoints(data),
            ["airac"] = string.IsNullOrEmpty(GetString(parameters, "airac")) ? null : GetString(parameters, "airac"),
        };
    }

    /// <summary>
    /// The actual planned route (origin -> destination only) as an
    /// ordered list of {lat, lon} - lets the map draw the real route
    /// instead of a straight line. Best-effort: a missing/malformed
    /// navlog just means an empty list, never fatal.
    /// </summary>
    private static JsonArray ParseRoutePoints(JsonObject data)
    {
        var points = new JsonArray();
        if (data["navlog"] is not JsonObject navlog || navlog["fix"] is not JsonArray fixes) return points;

        foreach (var fix in fixes)
        {
            if (fix is not JsonObject fixObj) continue;
            if (TryGetDouble(fixObj["pos_lat"], out var lat) && TryGetDouble(fixObj["pos_long"], out var lon))
            {
                points.Add(new JsonObject { ["lat"] = lat, ["lon"] = lon });
            }
        }
        return points;
    }

    /// <summary>
    /// Returns {ok:true, origin, destination, alternate, duration_minutes,
    /// aircraft_name, planned_at, route_points, airac} or
    /// {ok:false, error}. The error string is a translation key when
    /// recognized, else a raw message.
    /// </summary>
    public static async Task<JsonObject> FetchLatestOfp(string? username)
    {
        if (string.IsNullOrWhiteSpace(username))
        {
            return new JsonObject { ["ok"] = false, ["error"] = "simbrief_no_username" };
        }

        HttpResponseMessage response;
        try
        {
            var url = $"{SimbriefUrl}?username={Uri.EscapeDataString(username.Trim())}&json=1";
            response = await Http.GetAsync(url);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            return new JsonObject { ["ok"] = false, ["error"] = ex.Message };
        }

        if (!response.IsSuccessStatusCode)
        {
            return new JsonObject { ["ok"] = false, ["error"] = $"HTTP {(int)response.StatusCode}" };
        }

        JsonObject parsed;
        try
        {
            var body = await response.Content.ReadAsStringAsync();
            var data = JsonNode.Parse(body) as JsonObject ?? throw new FormatException("empty response");
            parsed = ParseOfp(data);
        }
        catch (Exception ex)
        {
            return new JsonObject { ["ok"] = false, ["error"] = $"unexpected response ({ex.Message})" };
        }

        parsed["ok"] = true;
        return parsed;
    }

    private static string GetRequiredIcao(JsonObject data, string field)
    {
        if (data[field] is not JsonObject obj || GetString(obj, "icao_code") is not { } icao || icao.Length == 0)
        {
            throw new KeyNotFoundException(field);
        }
        return icao;
    }

    private static string? GetString(JsonObject obj, string key) =>
        obj[key] is JsonValue value && value.TryGetValue<string>(out var s) ? s : null;

    private static bool TryGetLong(JsonNode? node, out long result)
    {
        result = 0;
        if (node is not JsonValue value) return false;
        if (value.TryGetValue<long>(out result)) return true;
        if (value.TryGetValue<string>(out var s) && long.TryParse(s, out result)) return true;
        return false;
    }

    private static bool TryGetDouble(JsonNode? node, out double result)
    {
        result = 0;
        if (node is not JsonValue value) return false;
        if (value.TryGetValue<double>(out result)) return true;
        if (value.TryGetValue<string>(out var s) && double.TryParse(s, System.Globalization.CultureInfo.InvariantCulture, out result)) return true;
        return false;
    }
}
